homeassistant/components/denonavr/media_player.py

```diff
@@ -74,58 +95,98 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
         )
+        self._available = True
+
+    def async_log_errors(  # pylint: disable=no-self-argument
+        func: Coroutine,
+    ) -> Coroutine:
+        """
+        Log errors occurred when calling a Denon AVR receiver.
+
+        Decorates methods of DenonDevice class.
+        Declaration of staticmethod for this method is at the end of this class.
+        """
+
+        @wraps(func)
+        async def wrapper(self, *args, **kwargs):
+            available = True
+            try:
+                return await func(self, *args, **kwargs)  # pylint: disable=not-callable
+            except AvrTimoutError:
+                available = False
+                if self._available is True:  # pylint: disable=protected-access
+                    _LOGGER.warning(
+                        "Timeout connecting to Denon AVR receiver at host %s. Device is unavailable",
+                        self._receiver.host,  # pylint: disable=protected-access
+                    )
+                    self._available = False  # pylint: disable=protected-access
+            except AvrForbiddenError:
+                available = False
+                if self._available is True:  # pylint: disable=protected-access
+                    _LOGGER.warning(
+                        "Denon AVR receiver at host %s responded with HTTP 403 error. Device is unavailable. Please consider power cycling your receiver",
+                        self._receiver.host,  # pylint: disable=protected-access
+                    )
+                    self._available = False  # pylint: disable=protected-access
+            except AvrCommandError as err:
+                _LOGGER.error(
+                    "Command %s failed with error: %s",
+                    func.__name__,  # pylint: disable=no-member
+                    err,
+                )
+            except DenonAvrError as err:
+                _LOGGER.error(
+                    "Error %s occurred in method %s for Denon AVR receiver",
+                    err,
+                    func.__name__,  # pylint: disable=no-member
+                    exc_info=True,
+                )
+            finally:
+                if (
+                    available is True
+                    and self._available is False  # pylint: disable=protected-access
+                ):
+                    _LOGGER.info(
+                        "Denon AVR receiver at host %s is available again",
+                        self._receiver.host,  # pylint: disable=protected-access
+                    )
+                    self._available = True  # pylint: disable=protected-access
+
+        return wrapper
 
     async def async_added_to_hass(self):
         """Register signal handler."""
         self.async_on_remove(
-            async_dispatcher_connect(self.hass, DOMAIN, self.signal_handler)
+            async_dispatcher_connect(self.hass, DOMAIN, self.async_signal_handler)
         )
```

Comment:
Can be removed once switched to entity services

Reply:
done